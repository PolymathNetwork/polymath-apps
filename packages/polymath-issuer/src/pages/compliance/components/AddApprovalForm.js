// @flow

import React, { Component, Fragment } from 'react';
import { connect } from 'react-redux';
import { Modal, Button } from '@polymathnetwork/ui';
import { withFormik } from 'formik';
import { Form, Dropdown, DropdownItem } from 'carbon-components-react';
import { RESTRICTION_TYPE } from '../../../constants';
import {
  bull,
  Box,
  PageCentered,
  ContentBox,
  Heading,
  FormItem,
  TextInput,
  RadioInput,
  PercentageInput,
  NumberInput,
  Grid,
  FormItemGroup,
  DatePickerInput,
  TimePickerSelect,
} from '@polymathnetwork/ui';
import validator from '@polymathnetwork/ui/validator';
import { addManualApproval } from '../../../actions/compliance';
import {
  validateTodayOrAfter,
  validateDays,
  validateExpiryTime,
  REQUIRED_MESSAGE,
  MORE_THAN_MESSAGE,
  ADDRESS_MESSAGE,
  MAX_DIGITS_MESSAGE,
} from '../../sto/components/ConfigureSTOForm/validators'; // this is mangled put into global object
import moment from 'moment';
import { toWei } from '../../../utils/contracts';

type Props = {
  isOpen: boolean,
  handleClose: () => any,
};

const formSchema = validator.object().shape({
  date: validator.object().shape({
    expiryDate: validator
      .date()
      .isRequired(REQUIRED_MESSAGE)
      .test('validateExpiryDate', validateTodayOrAfter),
    expiryTime: validator
      .number()
      .isRequired(REQUIRED_MESSAGE)
      .test('validateExpiryTime', validateExpiryTime),
  }),
  fromAddress: validator
    .string()
    .isRequired(REQUIRED_MESSAGE)
    .isAddress(ADDRESS_MESSAGE),
  toAddress: validator
    .string()
    .isRequired(REQUIRED_MESSAGE)
    .notOneOf(
      [validator.ref('fromAddress')],
      'To Address must be different than From Address'
    )
    .isAddress(ADDRESS_MESSAGE),
  token: validator
    .number()
    .moreThan(0, MORE_THAN_MESSAGE)
    .isRequired(REQUIRED_MESSAGE),
  description: validator.string().isRequired(REQUIRED_MESSAGE),
});

const initialValues = {
  date: {
    startDate: null,
    startTime: null,
  },
  fromAddress: '',
  toAddress: '',
  token: null,
  description: '',
};

export const AddApprovalComponent = ({
  handleSubmit,
  handleClose,
  handleChange,
  values,
  setFieldValue,
  errors,
  touched,
}) => {
  return (
    <Form className="global-restrictions" onSubmit={handleSubmit}>
      <Grid>
        <Grid.Row>
          <Grid.Col gridSpan={12}>
            <FormItem name="token">
              <FormItem.Label>
                <strong>Allow Transfer of</strong>
              </FormItem.Label>
              <FormItem.Input
                placeholder="Enter amount"
                min={1}
                component={NumberInput}
                unit="TOKENS"
              />
              <FormItem.Error />
            </FormItem>
          </Grid.Col>
        </Grid.Row>
        <Grid.Row>
          <Grid.Col gridSpan={12}>
            <FormItem name="fromAddress">
              <FormItem.Label>
                <strong>From Investor Wallet Address</strong>
              </FormItem.Label>
              <FormItem.Input
                component={TextInput}
                placeholder="Enter wallet address"
              />
              <FormItem.Error />
            </FormItem>
          </Grid.Col>
        </Grid.Row>
        <Grid.Row>
          <Grid.Col gridSpan={12}>
            <FormItem name="toAddress">
              <FormItem.Label>
                <strong>To Investor Wallet Address</strong>
              </FormItem.Label>
              <FormItem.Input
                component={TextInput}
                placeholder="Enter wallet Address"
              />
              <FormItem.Error />
            </FormItem>
          </Grid.Col>
        </Grid.Row>
        <Grid.Row>
          <Grid.Col gridSpan={12}>
            <FormItemGroup>
              <FormItemGroup.Items>
                <FormItem name="date.expiryDate">
                  <FormItem.Label>Expiry Date</FormItem.Label>
                  <FormItem.Input
                    component={DatePickerInput}
                    placeholder="mm / dd / yyyy"
                  />
                </FormItem>
                <FormItem name="date.expiryTime">
                  <FormItem.Label>Time</FormItem.Label>
                  <FormItem.Input
                    component={TimePickerSelect}
                    placeholder="hh:mm"
                  />
                </FormItem>
              </FormItemGroup.Items>
              <FormItemGroup.Error
                name="date"
                errors={errors}
                touched={touched}
              />
            </FormItemGroup>
          </Grid.Col>
        </Grid.Row>
        <Grid.Row>
          <Grid.Col gridSpan={12}>
            <FormItem name="description">
              <FormItem.Label>
                <strong>Description</strong>
              </FormItem.Label>
              <FormItem.Input
                component={TextInput}
                maxLength={32}
                placeholder="Start text here(up to 32 characters)"
              />
              <FormItem.Error />
            </FormItem>
          </Grid.Col>
        </Grid.Row>
      </Grid>
      <Modal.Footer>
        <Button kind="secondary" onClick={handleClose}>
          Cancel
        </Button>
        <Button type="submit">Confirm</Button>
      </Modal.Footer>
    </Form>
  );
};

const formikEnhancer = withFormik({
  validationSchema: formSchema,
  displayName: 'AddApprovalForm',
  validateOnChange: false,
  mapPropsToValues: props => {
    return {
      ...initialValues,
    };
  },
  handleSubmit: (values, { errors, setFieldError, props }) => {
    const { dispatch, handleClose, approvals } = props;
    const approvalExists = approvals.find(
      i =>
        i.fromAddress === values.fromAddress &&
        i.toAddress === values.toAddress &&
        i.expiry * 1000 > Date.now()
    );

    if (approvalExists) {
      setFieldError(
        'toAddress',
        'An approval with both addresses already exists'
      );
      setFieldError(
        'fromAddress',
        'An approval with both addresses already exists'
      );
      return;
    }
    const startsAt =
      moment(values.date.expiryDate).unix() * 1000 + values.date.expiryTime;

    dispatch(
      addManualApproval(
        values.fromAddress,
        values.toAddress,
        toWei(values.token),
        startsAt,
        values.description
      )
    );

    handleClose();
  },
});

const mapStateToProps = state => ({
  approvals: state.whitelist.approvals,
});

const FormikEnhancedForm = formikEnhancer(AddApprovalComponent);
const ConnectedForm = connect(mapStateToProps)(FormikEnhancedForm);

export default ConnectedForm;
