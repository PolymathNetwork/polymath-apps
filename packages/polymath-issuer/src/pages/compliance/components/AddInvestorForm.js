// @flow

import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Field, reduxForm, formValueSelector } from 'redux-form';
import { Form, Button } from 'carbon-components-react';
import { Remark } from '@polymathnetwork/ui';
import { TextInput, CheckboxInput } from '@polymathnetwork/ui/deprecated';
import { required, ethereumAddress } from '@polymathnetwork/ui/validate';

import LockupDatesFields, { validate } from './LockupDatesFields';
import type { RootState } from '../../../redux/reducer';

type StateProps = {|
  isSalePermanent: boolean,
  isPurchasePermanent: boolean,
  isPercentagePaused: boolean,
|};

type Props = {|
  handleSubmit: () => void,
  onClose: () => void,
|} & StateProps;

export const formName = 'add_investor_form';

const formValue = formValueSelector(formName);

const mapStateToProps = (state: RootState) => ({
  isSalePermanent: !!formValue(state, 'permanentSale'),
  isPurchasePermanent: !!formValue(state, 'permanentPurchase'),
  isPercentagePaused: state.whitelist.percentageTM.isPaused,
});

class AddInvestorForm extends Component<Props> {
  render() {
    return (
      <Form onSubmit={this.props.handleSubmit}>
        <Field
          name="address"
          component={TextInput}
          label="Investor&apos;s ETH Address"
          placeholder="Enter Investor&apos;s ETH Address"
          validate={[required, ethereumAddress]}
        />
        <LockupDatesFields
          isSalePermanent={this.props.isSalePermanent}
          isPurchasePermanent={this.props.isPurchasePermanent}
        />
        {!this.props.isPercentagePaused ? (
          <div>
            <br />
            <Field
              name="isPercentage"
              component={CheckboxInput}
              label="Exempt from % ownership restriction"
            />
          </div>
        ) : (
          ''
        )}
        <br />
        <Remark title="Reminder">
          Investors must be approved before they are added to the whitelist.
        </Remark>
        <p align="right">
          <Button kind="secondary" onClick={this.props.onClose}>
            Cancel
          </Button>
          <Button type="submit">Add New Investor</Button>
        </p>
      </Form>
    );
  }
}

export default connect(mapStateToProps)(
  reduxForm({
    form: formName,
    validate,
  })(AddInvestorForm)
);
