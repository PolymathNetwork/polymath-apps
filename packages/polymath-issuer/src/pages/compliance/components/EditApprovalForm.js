// @flow

import React, { Component, Fragment } from 'react';
import { connect } from 'react-redux';
import { Modal, Button } from '@polymathnetwork/ui';
import { withFormik } from 'formik';
import { Form, Dropdown, DropdownItem } from 'carbon-components-react';
import { RESTRICTION_TYPE } from '../../../constants';
import {
  bull,
  Box,
  PageCentered,
  ContentBox,
  Heading,
  FormItem,
  TextInput,
  RadioInput,
  PercentageInput,
  NumberInput,
  Grid,
  FormItemGroup,
  DatePickerInput,
  TimePickerSelect,
} from '@polymathnetwork/ui';
import validator from '@polymathnetwork/ui/validator';
import { editManualApproval } from '../../../actions/compliance';
import {
  validateTodayOrAfter,
  validateDays,
  validateExpiryTime,
  REQUIRED_MESSAGE,
  MORE_THAN_MESSAGE,
  ADDRESS_MESSAGE,
  MAX_DIGITS_MESSAGE,
} from '../../sto/components/ConfigureSTOForm/validators'; // this is mangled put into global object
import moment from 'moment';
import { toWei } from '../../../utils/contracts';

type Props = {
  isOpen: boolean,
  handleClose: () => any,
};

const formSchema = validator.object().shape({
  date: validator.object().shape({
    expiryDate: validator
      .date()
      .isRequired(REQUIRED_MESSAGE)
      .test('validateExpiryDate', validateTodayOrAfter),
    expiryTime: validator
      .number()
      .isRequired(REQUIRED_MESSAGE)
      .test('validateExpiryTime', validateExpiryTime),
  }),
  fromAddress: validator
    .string()
    .isRequired(REQUIRED_MESSAGE)
    .isAddress(ADDRESS_MESSAGE),
  toAddress: validator
    .string()
    .isRequired(REQUIRED_MESSAGE)
    .isAddress(ADDRESS_MESSAGE),
  token: validator
    .number()
    .moreThan(0, MORE_THAN_MESSAGE)
    .isRequired(REQUIRED_MESSAGE),
  description: validator.string().isRequired(REQUIRED_MESSAGE),
});

const initialValues = {
  date: {
    startDate: null,
    startTime: null,
  },
  fromAddress: '',
  toAddress: '',
  token: null,
  description: '',
};

export const AddApprovalComponent = ({
  handleSubmit,
  handleClose,
  handleChange,
  values,
  setFieldValue,
  errors,
  touched,
}) => {
  return (
    <Form className="global-restrictions" onSubmit={handleSubmit}>
      <Grid>
        <Grid.Row>
          <Grid.Col gridSpan={12}>
            <FormItem name="token">
              <FormItem.Label>
                <strong>Allow Transfer of</strong>
              </FormItem.Label>
              <FormItem.Input
                placeholder="Enter amount"
                min={1}
                component={NumberInput}
                unit="TOKENS"
              />
              <FormItem.Error />
            </FormItem>
          </Grid.Col>
        </Grid.Row>
        <Grid.Row>
          <Grid.Col gridSpan={12}>
            <FormItem name="fromAddress">
              <FormItem.Label>
                <strong>From Investor Wallet Address</strong>
              </FormItem.Label>
              <FormItem.Input
                component={TextInput}
                disabled={true}
                placeholder="Enter wallet address"
              />
              <FormItem.Error />
            </FormItem>
          </Grid.Col>
        </Grid.Row>
        <Grid.Row>
          <Grid.Col gridSpan={12}>
            <FormItem name="toAddress">
              <FormItem.Label>
                <strong>To Investor Wallet Address</strong>
              </FormItem.Label>
              <FormItem.Input
                component={TextInput}
                disabled={true}
                placeholder="Enter wallet Address"
              />
              <FormItem.Error />
            </FormItem>
          </Grid.Col>
        </Grid.Row>
        <Grid.Row>
          <Grid.Col gridSpan={12}>
            <FormItemGroup>
              <FormItemGroup.Items>
                <FormItem name="date.expiryDate">
                  <FormItem.Label>Expiry Date</FormItem.Label>
                  <FormItem.Input
                    component={DatePickerInput}
                    placeholder="mm / dd / yyyy"
                  />
                </FormItem>
                <FormItem name="date.expiryTime">
                  <FormItem.Label>Time</FormItem.Label>
                  <FormItem.Input
                    component={TimePickerSelect}
                    placeholder="hh:mm"
                  />
                </FormItem>
              </FormItemGroup.Items>
              <FormItemGroup.Error
                name="date"
                errors={errors}
                touched={touched}
              />
            </FormItemGroup>
          </Grid.Col>
        </Grid.Row>
        <Grid.Row>
          <Grid.Col gridSpan={12}>
            <FormItem name="description">
              <FormItem.Label>
                <strong>Description</strong>
              </FormItem.Label>
              <FormItem.Input
                component={TextInput}
                maxLength={32}
                placeholder="Start text here(up to 32 characters)"
              />
              <FormItem.Error />
            </FormItem>
          </Grid.Col>
        </Grid.Row>
      </Grid>
      <Modal.Footer>
        <Button kind="secondary" onClick={handleClose}>
          Cancel
        </Button>
        <Button type="submit">Confirm</Button>
      </Modal.Footer>
    </Form>
  );
};

const formikEnhancer = withFormik({
  validationSchema: formSchema,
  displayName: 'AddApprovalForm',
  validateOnChange: false,
  mapPropsToValues: props => {
    const { approval } = props;
    let expiryStartTime =
      approval.expiry -
      moment
        .unix(approval.expiry)
        .startOf('day')
        .unix();
    return {
      date: {
        expiryDate: moment.unix(approval.expiry).startOf('day'),
        expiryTime: expiryStartTime * 1000,
      },
      fromAddress: approval.fromAddress,
      toAddress: approval.toAddress,
      token: parseFloat(approval.tokens),
      description: approval.description,
    };
  },
  handleSubmit: (values, { errors, setFieldError, props }) => {
    const { dispatch, handleClose } = props;
    const startsAt =
      moment(values.date.expiryDate).unix() * 1000 + values.date.expiryTime;

    dispatch(
      editManualApproval(
        values.fromAddress,
        values.toAddress,
        toWei(values.token),
        startsAt,
        values.description
      )
    );

    handleClose();
  },
});

const mapStateToProps = state => ({
  approval: state.whitelist.editingApproval,
});

const FormikEnhancedForm = formikEnhancer(AddApprovalComponent);
const ConnectedForm = connect(mapStateToProps)(FormikEnhancedForm);

export default ConnectedForm;
