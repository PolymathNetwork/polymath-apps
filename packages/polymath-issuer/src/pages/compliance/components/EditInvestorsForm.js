// @flow

import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Field, reduxForm, formValueSelector } from 'redux-form';
import { Form, Button } from 'carbon-components-react';
import { CheckboxInput } from '@polymathnetwork/ui/deprecated';

import LockupDatesFields, { validate } from './LockupDatesFields';

type StateProps = {|
  isSalePermanent: boolean,
  isPurchasePermanent: boolean,
  isPercentagePaused: boolean,
|};

type Props = {|
  handleSubmit: () => void,
  onClose: () => void,
|} & StateProps;

export const formName = 'edit_investors_form';

const formValue = formValueSelector(formName);

const mapStateToProps = state => ({
  isSalePermanent: !!formValue(state, 'e_permanentSale'),
  isPurchasePermanent: !!formValue(state, 'e_permanentPurchase'),
  isPercentagePaused: state.whitelist.percentageTM.isPaused,
});

class EditInvestorsForm extends Component<Props> {
  render() {
    return (
      <Form onSubmit={this.props.handleSubmit}>
        <LockupDatesFields
          edit
          isSalePermanent={this.props.isSalePermanent}
          isPurchasePermanent={this.props.isPurchasePermanent}
        />
        {!this.props.isPercentagePaused ? (
          <div>
            <br />
            <Field
              name="e_isPercentage"
              component={CheckboxInput}
              label="Exempt from % ownership restriction"
            />
          </div>
        ) : (
          ''
        )}
        <br />
        <p align="right">
          <Button kind="secondary" onClick={this.props.onClose}>
            Cancel
          </Button>
          <Button type="submit">Submit</Button>
        </p>
      </Form>
    );
  }
}

export default connect(mapStateToProps)(
  reduxForm({
    form: formName,
    validate,
  })(EditInvestorsForm)
);
