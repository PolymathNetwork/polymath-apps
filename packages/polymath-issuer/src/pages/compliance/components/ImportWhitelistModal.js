// @flow

import React, { Component } from 'react';
import { connect } from 'react-redux';
import {
  Icon,
  FileUploader,
  Button,
  InlineNotification,
} from 'carbon-components-react';
import { Modal, Remark, Paragraph } from '@polymathnetwork/ui';
import { uploadCSV, resetUploaded } from '../../../actions/compliance';

import { fetch } from '../../../actions/sto';
import { STAGE_OVERVIEW } from '../../../reducers/sto';

import type { RootState } from '../../../redux/reducer';

type StateProps = {|
  isTooMany: boolean,
  parseError: String,
  isReady: boolean,
  isInvalid: boolean,
|};

type DispatchProps = {|
  fetch: () => any,
  uploadCSV: (file: Object) => any,
  resetUploaded: () => any,
|};

type Props = {|
  isOpen: boolean,
  onClose: () => any,
  onSubmit: (file: ?Object) => any,
|} & StateProps &
  DispatchProps;

const mapStateToProps = (state: RootState) => ({
  isTooMany: state.whitelist.isTooMany,
  parseError: state.whitelist.parseError,
  isReady: state.whitelist.uploaded.length > 0,
  isInvalid: state.whitelist.criticals.length > 0,
  sto: state.sto,
});

const mapDispatchToProps = {
  fetch,
  uploadCSV,
  resetUploaded,
};

class ImportWhitelistModal extends Component<Props> {
  componentDidMount() {
    this.props.fetch();
  }

  handleClose = () => {
    this.fileUploader.clearFiles();
    this.props.onClose();
  };

  handleSubmit = () => {
    this.handleClose();
    this.props.onSubmit();
  };

  handleUploaded = async (event: Object) => {
    const file = event.target.files[0];
    if (file.type.match(/csv.*/) || file.name.match(/.*\.csv$/i)) {
      await this.props.uploadCSV(file);
      //NOTE @sajclarke: This hack is necessary to add an eventlistener to the dynamic filename container from FileUploader
      const node = this.fileUploader.nodes[0];
      if (node) {
        const el = Array.from(node.getElementsByClassName('bx--file-close'))[0];
        el.addEventListener(
          'click',
          e => {
            this.props.resetUploaded();
          },
          false
        );
      }
    }
  };

  fileUploaderRef = (el: ?Object) => {
    // $FlowFixMe
    this.fileUploader = el;
  };

  render() {
    const {
      isOpen,
      sto,
      isTooMany,
      parseError,
      isReady,
      isInvalid,
    } = this.props;

    return (
      <Modal
        isOpen={isOpen}
        onClose={this.handleClose}
        className="whitelist-import-modal"
      >
        <Modal.Header>Import Whitelist</Modal.Header>
        <h4 className="pui-h4">
          Add multiple addresses to the whitelist by uploading a comma separated
          .CSV file.{' '}
          <strong>You may add up to 40 addresses per .CSV file</strong>. The
          format should be as follows:
          <br />• ETH Address (address to whitelist);
          <br />• Sell Restriction Date: <strong>mm/dd/yyyy</strong> (date when
          the resale restrictions should be lifted for that address);
          <br />
          Empty cell will be considered as permanent lockup.
          <br />• Buy Restriction Date: <strong> mm/dd/yyyy</strong> (date when
          the buy restrictions should be lifted for that address);
          <br />• KYC/AML Expiry Date: <strong>mm/dd/yyyy</strong>;<br />• Can
          buy from STO: set to <strong>"TRUE"</strong> to allow this address to
          purchase any number of tokens OR leave an empty cell to disable;
          <br />• Exempt From % Ownership: <strong>"TRUE"</strong> to enable OR
          empty cell to disable;
          <br />
          <br />
          {sto.stage === STAGE_OVERVIEW &&
          sto.details.type === 'USDTieredSTO' ? (
            <p>
              If you have scheduled a USD Tiered STO, please include the
              additional fields:
              <br />• Is Accredited: Set to <strong>"TRUE"</strong> to mark the
              address as that of an accredited investor OR leave empty to mark
              the address as that of a non-accredited investor
              <br />• Non Accredited Limit: Set a maximum investment limit for
              that non-accredited investor's address or leave empty to use the
              default limit programmed in the STO
              <br />
              <br />
              Important:
              <br />
              Is Accredited and Non Accredited Limit will be ignored if you have
              not yet scheduled your USD Tiered STO. If you have scheduled your
              STO, all accredited/non-accredited investor information will be
              imported adequately. If you have not, you will be required to
              re-upload this information.
              <br />
              <br /> Maximum numbers of investors per transaction is{' '}
              <strong>75</strong>.
            </p>
          ) : null}
        </h4>
        <h5 className="pui-h5">
          You can&nbsp;&nbsp;&nbsp;
          <Icon name="download" fill="#252D6B" width="16" height="16" />
          &nbsp;
          <a href="/whitelist-sample.csv" className="pui-bold" download>
            Download Sample.csv
          </a>
          &nbsp;&nbsp;file and edit it.
        </h5>
        <FileUploader
          iconDescription="Cancel"
          buttonLabel="Upload File"
          onChange={this.handleUploaded}
          className="file-uploader"
          accept={['.csv']}
          buttonKind="secondary"
          filenameStatus="edit"
          ref={this.fileUploaderRef}
        />
        {parseError.length > 0 ? (
          <InlineNotification
            hideCloseButton
            title={parseError}
            subtitle="Please check instructions above and try again."
            kind="error"
          />
        ) : isInvalid ? (
          <InlineNotification
            hideCloseButton
            title="The file you uploaded has invalid fields"
            subtitle="Please check instructions above and try again."
            kind="error"
          />
        ) : isInvalid && !isReady ? (
          <InlineNotification
            hideCloseButton
            title="The file you uploaded does not contain any valid values"
            subtitle="Please check instructions above and try again."
            kind="error"
          />
        ) : isTooMany ? (
          <InlineNotification
            hideCloseButton
            title="The file you uploaded contains more than 40 investors"
            subtitle="You can still continue, but only 40 first investors will be submitted."
            kind="error"
          />
        ) : (
          <Remark title="Reminder">
            Investors must be approved before they are added to the whitelist.{' '}
            <br />
            Your file cannot exceed 75 addresses. If you have more than 75
            addresses on your whitelist, upload multiple files.
          </Remark>
        )}
        <Modal.Footer>
          <Paragraph align="right">
            <Button
              className="cancel-btn"
              kind="secondary"
              onClick={this.handleClose}
            >
              Cancel
            </Button>
            <Button
              type="submit"
              kind="primary"
              disabled={!isReady || isInvalid}
              onClick={this.handleSubmit}
            >
              Import Whitelist
            </Button>
          </Paragraph>
        </Modal.Footer>
      </Modal>
    );
  }
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ImportWhitelistModal);
