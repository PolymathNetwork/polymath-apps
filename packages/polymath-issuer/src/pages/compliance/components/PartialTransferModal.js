// @flow

import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Modal, Button } from '@polymathnetwork/ui';
import { withFormik } from 'formik';
import { Form } from 'carbon-components-react';
import {
  bull,
  PageCentered,
  ContentBox,
  Heading,
  FormItem,
  TextInput,
} from '@polymathnetwork/ui';
import validator from '@polymathnetwork/ui/validator';
import { addAddressToPartialExempt } from '../../../actions/compliance';

type Props = {
  isOpen: boolean,
  handleClose: () => any,
};

const formSchema = validator.object().shape({
  address: validator
    .string()
    .isAddress('Invalid Address')
    .isRequired('Required'),
});

export const AddPartialTransferComponent = ({ handleSubmit, handleClose }) => (
  <Form onSubmit={handleSubmit}>
    <FormItem name="address">
      <Heading className="form-item-header" variant="h3">
        Exempted Wallet Address
      </Heading>
      <FormItem.Input component={TextInput} />
      <FormItem.Error />
    </FormItem>

    <Modal.Footer>
      <Button onClick={handleClose} className="cancel-btn" kind="secondary">
        Cancel
      </Button>
      <Button type="submit">Confirm</Button>
    </Modal.Footer>
  </Form>
);

const formikEnhancer = withFormik({
  validationSchema: formSchema,
  displayName: 'PartialAddressForm',
  validateOnChange: false,
  handleSubmit: (values, { setFieldError, props }) => {
    const { dispatch, partialAddresses } = props;
    const addressExists = partialAddresses.find(
      i => i.address === values.address
    );
    if (addressExists) {
      setFieldError('address', 'Address is already an Exempted Address');
      return;
    }
    props.handleClose();
    dispatch(addAddressToPartialExempt(values.address));
  },
});

const mapStateToProps = state => ({
  partialAddresses: state.whitelist.partialAddresses,
});

const FormikEnhancedForm = formikEnhancer(AddPartialTransferComponent);
const ConnectedForm = connect(mapStateToProps)(FormikEnhancedForm);

export class PartialTransferModal extends Component<Props> {
  render() {
    const { isOpen, handleClose } = this.props;
    return (
      <Modal isOpen={isOpen} onClose={handleClose}>
        <Modal.Header variant="alert">
          Add an address to exempt a wallet from the Restricted Partial
          Transfers rules
        </Modal.Header>
        <Modal.Body>
          <p>
            Enter the wallet address you wish to exempt from the Restricted
            Partial Transfers rule. This wallet will be able to trade freely
            (unless restricted by other rules). Consult with your legal team
            before adding a new wallet to the list.
          </p>
          <ConnectedForm handleClose={handleClose} />
        </Modal.Body>
      </Modal>
    );
  }
}

export default PartialTransferModal;
