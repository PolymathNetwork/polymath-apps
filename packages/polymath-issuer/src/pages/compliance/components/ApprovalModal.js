// @flow

import React, { Component } from 'react';
import { connect } from 'react-redux';
import { withFormik } from 'formik';
import { Form } from 'carbon-components-react';
import { Remark, Modal, Button } from '@polymathnetwork/ui';
import validator from '@polymathnetwork/ui/validator';
import { addAddressToTransferManager } from '../../../actions/compliance';
import AddApprovalForm from './AddApprovalForm';
import EditApprovalForm from './EditApprovalForm';

type Props = {
  isOpen: boolean,
  isEdit: boolean,
  handleClose: () => any,
};

class ApprovalModal extends Component<Props> {
  render() {
    const { isOpen, handleClose, isEdit } = this.props;
    return (
      <Modal isOpen={isOpen} onClose={handleClose}>
        <Modal.Body>
          <h1 className="pui-h2">{`${
            isEdit ? 'Edit' : 'Create'
          } Trade Approval`}</h1>
          {isEdit ? (
            <EditApprovalForm handleClose={handleClose} />
          ) : (
            <AddApprovalForm handleClose={handleClose} />
          )}
        </Modal.Body>
      </Modal>
    );
  }
}

export default ApprovalModal;
