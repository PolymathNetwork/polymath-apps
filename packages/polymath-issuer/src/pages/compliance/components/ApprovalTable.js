// @flow

import React, { Component } from 'react';
import { DataTable, Icon } from 'carbon-components-react';
import { Button } from '@polymathnetwork/ui';
import ApprovalModal from './ApprovalModal';
import { connect } from 'react-redux';
import { removeApprovalFromApprovals } from '../../../actions/compliance';
import moment from 'moment';
const {
  TableContainer,
  Table,
  TableHead,
  TableRow,
  TableBody,
  TableCell,
  TableHeader,
  TableToolbar,
  TableToolbarContent,
} = DataTable;

const columns = [
  {
    header: 'From Investor Wallet Address',
    key: 'fromAddress',
  },
  {
    header: 'To Investor Wallet Address',
    key: 'toAddress',
  },
  {
    header: 'Approval Expiry',
    key: 'expiry',
  },
  // {
  //   header: 'Unique ID (TxHash) of Approval',
  //   key: 'txhash',
  // },
  {
    header: 'Description',
    key: 'description',
  },
  {
    header: 'Approved Token Transfer',
    key: 'tokens',
  },
  {
    header: 'Tokens Transferred to Date',
    key: 'tokensTransferred',
  },
];

const emptyRow = [
  {
    id: '0',
    fromAddress: '-',
    toAddress: '-',
    // txhash: '-',
    description: '-',
    tokens: '-',
    tokensTransferred: '-',
  },
];

type State = {|
  isApprovalModalOpen: boolean,
|};

class ApprovalTable extends Component<Props, State> {
  state = {
    isApprovalModalOpen: false,
  };

  handleOpen = () => {
    this.setState({ isApprovalModalOpen: true });
  };

  handleClose = () => {
    this.setState({ isApprovalModalOpen: false });
  };

  handleDelete = id => {
    this.props.removeApprovalFromApprovals(id);
  };

  formatCell = cell => {
    if (!cell.value) return '-';
    switch (cell.info.header) {
      case 'expiry':
        return moment.unix(cell.value).format('MMM DD YYYY');
      default:
        return cell.value;
    }
  };
  render() {
    const { approvals } = this.props;
    return (
      <div>
        <ApprovalModal
          isOpen={this.state.isApprovalModalOpen}
          handleClose={this.handleClose}
        />
        <DataTable
          headers={columns}
          rows={approvals < 1 ? emptyRow : approvals}
          render={({ rows, headers, getHeaderProps }) => {
            return (
              <TableContainer title="">
                <TableToolbar>
                  <TableToolbarContent>
                    {/* pass in `onInputChange` change here to make filtering work */}
                    <Button onClick={this.handleOpen} icon="icon--add">
                      Add new exemption
                    </Button>
                  </TableToolbarContent>
                </TableToolbar>
                <Table>
                  <TableHead>
                    <TableRow>
                      {headers.map(header => (
                        <TableHeader
                          key={header}
                          {...getHeaderProps({ header })}
                        >
                          {header.header}
                        </TableHeader>
                      ))}
                      <TableHeader />
                    </TableRow>
                  </TableHead>
                  <TableBody>
                    {rows.map(row => (
                      <TableRow
                        key={row.id}
                        onMouseOver={() => console.log('test')}
                      >
                        {row.cells.map(cell => (
                          <TableCell key={cell.id}>
                            {this.formatCell(cell)}
                          </TableCell>
                        ))}
                        {approvals.length > 0 ? (
                          <TableCell>
                            <div style={{ display: 'flex' }}>
                              <Icon
                                className="table-icon"
                                name="edit"
                                width="12"
                                height="12"
                              />
                              <Icon
                                onClick={() => this.handleDelete(row.id)}
                                className="table-icon"
                                style={{ marginLeft: '10px' }}
                                name="delete"
                                width="12"
                                height="12"
                              />
                            </div>
                          </TableCell>
                        ) : (
                          <TableCell />
                        )}
                      </TableRow>
                    ))}
                  </TableBody>
                </Table>
              </TableContainer>
            );
          }}
        />
      </div>
    );
  }
}

const mapStateToProps = state => ({
  approvals: state.whitelist.approvals,
});

const mapDispatchToProps = {
  removeApprovalFromApprovals,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ApprovalTable);
