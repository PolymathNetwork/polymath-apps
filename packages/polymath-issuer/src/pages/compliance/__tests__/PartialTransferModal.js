import { PartialTransferModal } from '../components/PartialTransferModal';
import DefaultPartialTransferModal from '../components/PartialTransferModal';
import renderer from 'react-test-renderer';
import React from 'react';
import configureStore from 'redux-mock-store';
import { theme, GlobalStyles } from '@polymathnetwork/ui';
import { shallow, mount } from 'enzyme';
import { ThemeProvider } from 'styled-components';
import moment from 'moment-timezone';
import { Provider } from 'react-redux';
const mockStore = configureStore([]);
moment.tz.setDefault('America/New_York');

describe('Partial Transfer Modal', () => {
  it('should render correctly', () => {
    const props = {
      fetchWhitelist: jest.fn(),
      fetchManagers: jest.fn(),
      fetchPartialTransfers: jest.fn(),
      theme,
    };
    const tree = shallow(<PartialTransferModal {...props} />);
    expect(tree).toMatchSnapshot();
  });
});
