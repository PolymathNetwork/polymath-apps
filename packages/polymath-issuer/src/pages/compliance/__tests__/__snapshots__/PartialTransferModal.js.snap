// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Partial Transfer Modal should render correctly 1`] = `
<Modal__StyledModal>
  <Header
    status="idle"
    variant="alert"
  >
    Add an address to exempt a wallet from the Restricted Partial Transfers rules
  </Header>
  <Body>
    <p>
      Enter the wallet address you wish to exempt from the Restricted Partial Transfers rule. This wallet will be able to trade freely (unless restricted by other rules). Consult with your legal team before adding a new wallet to the list.
    </p>
    <Connect(WithFormik(AddPartialTransferComponent)) />
  </Body>
</Modal__StyledModal>
`;
